cmd/nitro/nitro.go

```diff
@@ -55,6 +58,242 @@ func printSampleUsage(name string) {
+			account.Address = common.HexToAddress(walletConf.Account)
+			account, err = myKeystore.Find(account)
+		} else {
+			if walletConf.Password() == nil {
+				return common.Address{}, errors.New("l2 password not set")
+			}
+			if devPrivKey == nil {
+				return common.Address{}, errors.New("l2 private key not set")
+			}
+			account, err = myKeystore.ImportECDSA(devPrivKey, *walletConf.Password())
+		}
+		if err != nil {
+			return common.Address{}, err
+		}
+		if walletConf.Password() == nil {
+			return common.Address{}, errors.New("l2 password not set")
+		}
+		err = myKeystore.Unlock(account, *walletConf.Password())
+		if err != nil {
+			return common.Address{}, err
+		}
+	}
+	return devAddr, nil
+}
+
+func downloadInit(initConfig *InitConfig) string {
+	if initConfig.Url == "" {
+		return ""
+	}
+	if strings.HasPrefix(initConfig.Url, "file:") {
+		return initConfig.Url[5:]
+	}
+	grabclient := grab.NewClient()
+	log.Info("Downloading initial database", "url", initConfig.Url)
+	fmt.Println()
+	printTicker := time.NewTicker(time.Second)
+	defer printTicker.Stop()
+	attempt := 0
+	for {
+		attempt++
+		req, err := grab.NewRequest(initConfig.DownloadPath, initConfig.Url)
+		if err != nil {
+			panic(err)
+		}
+		resp := grabclient.Do(req)
+		firstPrintTime := time.Now().Add(time.Second * 2)
+	updateLoop:
+		for {
+			select {
+			case <-printTicker.C:
+				if time.Now().After(firstPrintTime) {
+					bps := resp.BytesPerSecond()
+					done := resp.BytesComplete()
+					total := resp.Size()
+					timeRemaining := (time.Second * time.Duration(total-done)) / time.Duration(bps)
+					timeRemaining = timeRemaining.Truncate(time.Millisecond * 10)
+					fmt.Printf("\033[2K\r  transferred %v / %v bytes (%.2f%%) [%.2fMbps, %s remaining]",
+						done,
+						total,
+						resp.Progress()*100,
+						bps*8/1000000,
+						timeRemaining.String())
+				}
+			case <-resp.Done:
+				if err := resp.Err(); err != nil {
+					fmt.Printf("\033[2K\r  attempt %d failed: %v", attempt, err)
+					break updateLoop
+				}
+				log.Info("Download done", "filename", resp.Filename, "duration", resp.Duration())
+				fmt.Println()
+				return resp.Filename
+			}
+		}
+		<-time.After(initConfig.DownloadPoll)
+	}
+}
+
+func openInitializeChainDb(stack *node.Node, initConfig *InitConfig, chainId *big.Int, cacheConfig *core.CacheConfig) (ethdb.Database, *core.BlockChain, error) {
+	if !initConfig.Force {
+		if readOnlyDb, err := stack.OpenDatabaseWithFreezer("l2chaindata", 0, 0, "", "", true); err == nil {
+			if chainConfig := arbnode.TryReadStoredChainConfig(readOnlyDb); chainConfig != nil {
+				readOnlyDb.Close()
+				chainDb, err := stack.OpenDatabaseWithFreezer("l2chaindata", 0, 0, "", "", false)
+				if err != nil {
+					return chainDb, nil, err
+				}
+				l2BlockChain, err := arbnode.GetBlockChain(chainDb, cacheConfig, chainConfig)
+				return chainDb, l2BlockChain, err
```

Comment:
The "Check that this ArbOS state has the correct chain ID" check below should probably be moved here